### ISSUE TITLE
[BUG] The producer flush opertion is not guarantee to flush all messages

### ISSUE THREAD

**Issue by RobertIndie** [apache/pulsar-client-go#1057]
#### Expected behavior

The `producer.Flush` should guarantee that all the messages should be flushed to the broker.

#### Actual behavior

The flush operation may skip some messages in the producer.

#### Steps to reproduce

```go
producer.SendAsync(context.Background(), &ProducerMessage{
		Payload: []byte("hello"),
	}, nil)
err = producer.Flush()
if err != nil {
	return
}
```
Currently, the Flush cannot guarantee that the message `hello` is actually flushed to the broker.

The root cause is that when calling the SendAsync, the producer will create another go routine to put the message to the queue.
Therefore, when flush is called, the message may not have entered the queue yet.

#### System configuration
**Pulsar Go version**: master

**Comment by gunli** [comment]
@RobertIndie I have reported a bug in #1042 about this, and @zengguan is prepareing a PR to fix it.

**Comment by Gleiphir2769** [comment]
Hi @RobertIndie, good cache! It seems like a long standing problem. `producer.Flush` forget to check and clear `dataChan` .

https://github.com/apache/pulsar-client-go/blob/3812c07a0b8f6c37a2803cd740f4bf917160fd22/pulsar/producer_partition.go#L1302-L1312

If you have no time on this thread, I think I can fix it.

**Comment by gunli** [comment]
@Gleiphir2769 @RobertIndie Flush() just flush the message in the pending queue..., I thought we intentionally designed it this way, so in #1042 I just mentioned that Close() should consume the message int dataChan... 

> Now we should clear dataChan and put all the messages to internalSent before internalFlush.

I think the root cause is not the dataChan, pushing them into one channal will restore another bug #687, besides,  send message is a frequently call, use a explicit type channel(chan *sendRequest) is better than a generic type channel(chan interface{}).

I think we can improve it by these steps:
1. move the shema encoding, compressing logic from `internalSend` back to `internalSendAsync` to make it like a JAVA client did; 
2. consume the message in dataChan when we call Flush() and Close(). BTW, in my opinion, Flush() is useless, 'cause in Flush() we don't update the state of the producer, when we are in flushing, user can still call Send() to put new message into the chan, unless we allow user to do that; 
3. refactor the resource manage logic(request.done() we discussed before).

**Comment by Gleiphir2769** [comment]
>  I thought we intentionally designed it this way
Yes, you're right. Because in the origin design there is only one channel `eventChan` to handle both `send` and `flush` cmd. Which means when `flush` cmd arrived, all the `send` has be handled by `internalSend` and send to `pendingQueue`.   After #1029 , `eventChan` has been split into `dataChan` and `cmdChan` and the order guarantee has been broken.

**Comment by RobertIndie** [comment]
Thank all of you for figuring out the issue. @gunli @Gleiphir2769 

So looks like it's a regression bug introduced in https://github.com/apache/pulsar-client-go/pull/1029
